source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -222,10 +252,12 @@ void UdpProxyFilter::ActiveSession::write(const Buffer::Instance& buffer) {
 
   // NOTE: On the first write, a local ephemeral port is bound, and thus this write can fail due to
   //       port exhaustion.
-  // NOTE: We do not specify the local IP to use for the sendmsg call. We allow the OS to select
-  //       the right IP based on outbound routing rules.
+  // NOTE: We do not specify the local IP to use for the sendmsg call if use_original_src_ip_ is not
+  // set.
+  //       We allow the OS to select the right IP based on outbound routing rules.
+  const Network::Address::Ip* local_ip = use_original_src_ip_ ? addresses_.peer_->ip() : nullptr;
   Api::IoCallUint64Result rc =
-      Network::Utility::writeToSocket(*io_handle_, buffer, nullptr, *host_->address());
+      Network::Utility::writeToSocket(socket_->ioHandle(), buffer, local_ip, *host_->address());
```

Comment:
When I last inquired about this with @danzh2010 , I don't think this is going to preserve the downstream port, which may be a significant limitation. Will this preserve the port? If not, can we do it? If not, can we document the limitation?

Reply:
It does not preserve the port. I will document the limitation.